tests/federation_room_join_partial_state_test.go

```diff
@@ -1029,34 +1175,21 @@ func TestPartialStateJoin(t *testing.T) {
 		// wait until hs2 starts syncing state
 		fedStateIdsRequestReceivedWaiter.Waitf(t, 5*time.Second, "Waiting for /state_ids request")
 
-		syncResponseChan := make(chan gjson.Result)
-		go func() {
-			response, _ := charlie.MustSync(t, client.SyncReq{})
-			syncResponseChan <- response
-			close(syncResponseChan)
-		}()
+		response, nextBatch := charlie.MustSync(t, client.SyncReq{})
 
-		// the client-side requests should still be waiting
-		select {
-		case <-syncResponseChan:
-			t.Fatalf("hs2 sync completed before state resync complete")
-		default:
+		// the client-side requests shouldn't report the join yet
+		syncJoinedRoomPath := "rooms.join." + client.GjsonEscape(roomID)
+		if response.Get(syncJoinedRoomPath).Exists() {
+			t.Fatal("Sync shouldn't include the joined room yet")
 		}
 
 		// reply to hs2 with a bogus /state_ids response
 		fedStateIdsSendResponseWaiter.Finish()
 
-		// charlie's /sync request should now complete, with the new room
-		var syncRes gjson.Result
-		select {
-		case <-time.After(1 * time.Second):
-			t.Fatalf("hs2 /sync request request did not complete")
-		case syncRes = <-syncResponseChan:
-		}
+		response, _ = charlie.MustSync(t, client.SyncReq{Since: nextBatch})
 
-		roomRes := syncRes.Get("rooms.join." + client.GjsonEscape(roomID))
-		if !roomRes.Exists() {
-			t.Fatalf("hs2 /sync completed without join to new room\n")
+		if !response.Get(syncJoinedRoomPath).Exists() {
+			t.Fatal("hs2 /sync completed without join to new room")
 		}
```

Comment:
We could do with a comment here explaining that we expect hs2 to fall back to requesting state from hs1 and complete the partial state join.

Also syncs no longer block until joins complete, so we have to use `MustSyncUntil` or `awaitPartialStateJoinCompletion` to avoid creating more flakes.

Reply:
MustSyncUntil in 76d2e5c

Reply:
And comment in e1d3bbe